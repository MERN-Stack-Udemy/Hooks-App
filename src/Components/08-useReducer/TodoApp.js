import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer'
import { useForm } from '../../Hooks/useForm';

import './styles.css';

export const TodoApp = () => {

  const [ { description }, handleChage, reset ] = useForm({
    description:''
  });

  const init = () => {
    return JSON.parse( localStorage.getItem('todos') ) || [];
  }

  const [ todos, dispatch ] = useReducer(todoReducer, [], init);
  
  useEffect(() => {
    localStorage.setItem( 'todos', JSON.stringify(todos) );
  }, [todos])

  const handleSubmit = e => { 
    e.preventDefault();

    if( description.trim().length <=2){
      return;
    }
    
    const newToDo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    }

    const action = {
      type: 'ADD',
      payload: newToDo
    }
    
    dispatch( action );
    reset();
  }

  const handleToggle = ( todoId ) => {
    dispatch({
      type: 'TOGGLE',
      payload: todoId
    });
  }

  const handleDelete = todoId => {
    const action = {
      type: 'DELETE',
      payload: todoId
    };
    dispatch( action );
  } 

  return (
    <div>
      <h1>Todos ({todos.length})</h1>
      <hr/>

      <div className="row">
        <div className="col-7">
          <ul className="list-group list-group-flush">
            {
              todos.map((todo, i) => (
                <li 
                  key={todo.id}
                  className="list-group-item"
                >
                  <p
                    className={ `${ todo.done && 'complete'}`}
                    onClick={()=> handleToggle( todo.id )}
                  >{i+1}. {todo.desc}</p>
                  <button 
                    className="btn btn-danger"
                    onClick= { () => handleDelete(todo.id) }
                  >Delete</button>
                </li>
              ))
            }
          </ul>
        </div>
        <div className="col-5">

          <h5>Add ToDo</h5>
          <hr/>
          <form onSubmit={ handleSubmit }>
            <input
              autoComplete="off"
              className="form-control"
              name="description"
              onChange={ handleChage }
              placeholder="To Do..."
              type="text"
              value={ description }
            />
            
            <button
              className="btn btn-outline-primary mt-2 btn-block"
              type="submit"
            >
              Add
            </button>

          </form>

        </div>
      </div>


    </div>
  )
}
