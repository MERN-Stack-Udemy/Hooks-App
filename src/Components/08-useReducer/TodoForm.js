import React from "react";
import { useForm } from "../../Hooks/useForm";

export const TodoForm = ({ dispatch }) => {

  const [ { description }, handleChage, reset ] = useForm({
    description:''
  });

  const handleSubmit = e => { 
    e.preventDefault();

    if( description.trim().length <=2){
      return;
    }
    const newToDo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    }

    dispatch({
      type: 'ADD',
      payload: newToDo
    });
    reset();
  }
  
  return (
    <form onSubmit={handleSubmit}>
      <input
        autoComplete="off"
        className="form-control"
        name="description"
        onChange={handleChage}
        placeholder="To Do..."
        type="text"
        value={description}
      />

      <button className="btn btn-outline-primary mt-2 btn-block" type="submit">
        Add
      </button>
    </form>
  );
};
