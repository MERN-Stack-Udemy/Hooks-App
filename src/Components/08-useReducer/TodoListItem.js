import React from 'react'

export const TodoListItem = ({ todo, index, dispatch}) => {
  
  const handleToggle = ( todoId ) => {
    dispatch({
      type: 'TOGGLE',
      payload: todoId
    });
  }

  const handleDelete = todoId => {
    const action = {
      type: 'DELETE',
      payload: todoId
    };
    dispatch( action );
  } 

  return (
    <li
      className='list-group-item'
    >
      <p
        className={ `${ todo.done && 'complete'}`}
        onClick={()=> handleToggle( todo.id )}
      >{index+1}. {todo.desc}</p>
      <button 
        className="btn btn-danger"
        onClick= { () => handleDelete(todo.id) }
      >Delete</button>
    </li>
  )
}
