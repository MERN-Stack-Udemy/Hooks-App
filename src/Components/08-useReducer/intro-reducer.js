
const initialState = [{
  id: 1,
  todo: 'ir a correr',
  done: false
}];

const todoReducer = (state=initialState, action ) => {
  if(action?.type === 'ADD'){
    return [...initialState, action.payload ];
  }

  return state;
}

let todos = todoReducer();

const newTodo = {
  id: 2,
  todo: 'estudiar react',
  done: false
}

const addNewAction = {
  type: 'ADD',
  payload: newTodo,
  // dispatch
}

todos = todoReducer( todos, addNewAction );


console.log(todos); 