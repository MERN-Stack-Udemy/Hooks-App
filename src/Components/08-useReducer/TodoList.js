import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos, dispatch}) => {
  return (
    <ul className="list-group list-group-flush">
      {
        todos.map( (todo, index) => (
          <TodoListItem 
            dispatch={dispatch}
            index={index}
            key = {todo.id}
            todo={todo}
          />
        ) )
      }
    </ul>
  )
}
