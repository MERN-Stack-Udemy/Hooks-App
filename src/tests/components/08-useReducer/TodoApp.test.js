import { mount, shallow } from "enzyme"
import { act } from "react-dom/test-utils";
import { TodoApp } from "../../../Components/08-useReducer/TodoApp"
import { demoTodos } from "../fixtems/demoTodos";

describe('test on <TodoApp/>', () => {
    
    const wrapper = shallow(<TodoApp/>);
    Storage.prototype.setItem = jest.fn(() => {});

    test('should be match to snapshot', () => {
        expect( wrapper ).toMatchSnapshot();
    })
    
    test('should add a new TODO', () => {
        const wrapper = mount(<TodoApp/>);
        act(() => {
            wrapper.find('TodoForm').prop('dispatch')({
                type:'ADD',
                payload: demoTodos[0],
            });
            wrapper.find('TodoForm').prop('dispatch')({
                type:'ADD',
                payload: demoTodos[1],
            });
        })
        expect(wrapper.find('h1').text().trim() ).toEqual(`Todos (2)`)
        // when TodoForm add a new TODO, localStorage is called for save todos one by one
        expect( localStorage.setItem ).toHaveBeenCalledTimes(2);
    })
    
    test('should be delete one TODO', () => {
        wrapper.find('TodoForm').prop('dispatch')({ type: 'ADD', payload: demoTodos[0]});
        wrapper.find('TodoList').prop('dispatch')({ type: 'DELETE', payload: demoTodos[0].id});

        expect(wrapper.find('h1').text().trim() ).toEqual(`Todos (0)`);
    })
})
