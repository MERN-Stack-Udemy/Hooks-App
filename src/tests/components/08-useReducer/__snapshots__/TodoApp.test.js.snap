// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test on <TodoApp/> should be match to snapshot 1`] = `
<div>
  <h1>
    Todos (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        dispatch={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <h5>
        Add ToDo
      </h5>
      <hr />
      <TodoForm
        dispatch={[Function]}
      />
    </div>
  </div>
</div>
`;
